src/allocation_engine/interval_tree.rs

```diff
@@ -33,14 +37,117 @@ pub struct Constraint {
     pub policy: AllocPolicy,
 }
 
-/// A closed interval range [min, max] used to describe a
+/// A closed interval range [start, end] used to describe a
 /// memory slot that will be assigned to a device by the VMM.
 /// This structure represents the key of the Node object in
 /// the interval tree implementation.
-#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Hash, Ord)]
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Hash, Ord, Debug)]
 pub struct Range {
-    pub min: u64,
-    pub max: u64,
+    /// Lower boundary of the interval.
+    start: u64,
+    /// Upper boundary of the interval.
+    end: u64,
+}
+
+#[allow(clippy::len_without_is_empty)]
+impl Range {
+    /// Create a new Range object.
+    pub fn new(start: u64, end: u64) -> Result<Self> {
+        if start > end {
+            return Err(Error::InvalidRange(start, end));
+        }
+        Ok(Range { start, end })
+    }
+
+    /// Get length of the range.
+    pub fn len(&self) -> u64 {
+        self.end - self.start + 1
+    }
+
+    /// Check whether two Range objects overlap with each other.
+    pub fn overlaps(&self, other: &Range) -> bool {
+        max(self.start, other.start) <= min(self.end, other.end)
+    }
+
+    /// Check whether the key is fully covered.
+    pub fn contains(&self, other: &Range) -> bool {
+        self.start <= other.start && self.end >= other.end
+    }
+
+    /// Get the beginning of the range.
+    pub fn get_start(&self) -> u64 {
```

Comment:
Rust getters don't use the keyword `get`. You should write directly `start` and `end`.